### PR TITLE
mazda: add alpha longitudinal support

### PR DESCRIPTION
**Pull request by yummydirtx**
This pull request adds support for full alpha longitudinal on the Mazda CX-5. The method I used to achieve this involved poisoning the stock radar unit by placing it in "programming" session and sending "tester-present" messages to hold it in that state, and then essentially emulating that radar unit enough to prevent a full malfunction that would disable the ability to trigger MRCC. It is important to note that doing this necessarily disables AEB braking, and as of the current state of these changes, triggers a few malfunctions to show on the dash. None of these impair car function, and across many test drives, alpha longitudinal has worked without triggering errors or exhibiting any unusual unsafe behavior.

This is my first contribution to this project, and, clearly, it's a large one in size, so I welcome any and all critiques or suggestions that need to be addressed before merging all this.

The following is Copilot's summary of my changes (reviewed by me):

**Longitudinal Control Integration and Logic:**

* Added support for openpilot-managed longitudinal control, including logic for synthetic stop-and-go/HOLD, resume handling, and synthetic ACC message generation in `carcontroller.py`. This includes new state variables and careful management of transitions between stopping, holding, and starting, as well as handling both physical and virtual resume requests. [[1]](diffhunk://#diff-81fe04426a7c807332ac7cb3526d66b5275ded0729578fea78c1441181a58fe0L2-R21) [[2]](diffhunk://#diff-81fe04426a7c807332ac7cb3526d66b5275ded0729578fea78c1441181a58fe0R31-R39) [[3]](diffhunk://#diff-81fe04426a7c807332ac7cb3526d66b5275ded0729578fea78c1441181a58fe0R52-R53) [[4]](diffhunk://#diff-81fe04426a7c807332ac7cb3526d66b5275ded0729578fea78c1441181a58fe0R70-R208)

**Cruise State and Button Event Handling:**

* Refined parsing and reporting of cruise control button events (including main, cancel, accel, decel, and distance) to support both stock and openpilot longitudinal control. Updated cruise state detection to use radar signals when in openpilot longitudinal mode, ensuring accurate availability and engagement status. [[1]](diffhunk://#diff-be7075277f54e02ec59e95a15075be5a68095ba6175d64acbeee648258198128R24-R33) [[2]](diffhunk://#diff-be7075277f54e02ec59e95a15075be5a68095ba6175d64acbeee648258198128L83-R134) [[3]](diffhunk://#diff-be7075277f54e02ec59e95a15075be5a68095ba6175d64acbeee648258198128L115-L128)

**Car Interface and Parameters:**

* Updated `interface.py` to enable openpilot longitudinal control only for supported models (CX-5 2022), set appropriate car parameters (starting state, accel values, actuator delay, PID tuning), and configure the safety model accordingly. [[1]](diffhunk://#diff-e9f4ba4989d4b5ef60ca88571f8c260acba8aa2f26fbc5e6de2a0758110bbadeR7-R11) [[2]](diffhunk://#diff-e9f4ba4989d4b5ef60ca88571f8c260acba8aa2f26fbc5e6de2a0758110bbadeL17-R27) [[3]](diffhunk://#diff-e9f4ba4989d4b5ef60ca88571f8c260acba8aa2f26fbc5e6de2a0758110bbadeR40-R50)

**Radar Session Management:**

* Added initialization and deinitialization hooks to enter/exit radar programming session when openpilot longitudinal control is enabled, ensuring proper radar behavior during take-over and handoff.

These changes collectively enable a robust openpilot longitudinal control experience on supported Mazda vehicles, with careful handling of state transitions, user inputs, and compatibility with stock systems.

Validation
* Dongle ID: 21d103861daeed11
* Route: https://connect.comma.ai/21d103861daeed11/00000015--fba69e0004

### CODE REVIEW

File: opendbc/safety/modes/mazda.h [hunk 1]
```diff
@@ -70,6 +104,39 @@ static bool mazda_tx_hook(const CANPacket_t *msg) {
...
+      // opendbc/car/mazda/longitudinal.py. If this is tighter than the sender,
+      // Panda will silently drop 0x21b frames once ACCEL_CMD crosses the
+      // safety threshold, which looks like an unexplained set-speed unlatch.
+      const LongitudinalLimits MAZDA_LONG_LIMITS = {
+        .max_accel = 2000,
+        .min_accel = -2000,
+        .inactive_accel = 0,
+      };
+
+      int desired_accel = ((((int)msg->data[2] & 0x3U) << 11) | (((int)msg->data[3]) << 3) | (((int)msg->data[4]) >> 5)) - 4096;
```
[the remark this answers is not in the document]
Comment (yummydirtx) [reply opendbc/safety/modes/mazda.h:121]: I validated that this is an erroneous suggestion by packing CRZ_INFO.ACCEL_CMD through the repo’s DBC tooling and checking the decoded value path used in safety. In this stack, the 17|13@0+ field is encoded across data[2] low bits, all of data[3], and data[4] high bits, so the existing 3-byte extraction is the one that matches produced frames. The suggested 2-byte extraction decodes incorrect values for these packed messages. I kept the working extraction and added explicit Mazda longitudinal accel-limit test coverage to protect against regressions.

File: opendbc/safety/tests/test_mazda.py [hunk 2]
```diff
@@ -81,5 +81,20 @@ def test_buttons(self):
...
+  def setUp(self):
+    self.packer = CANPackerSafety("mazda_2017")
+    self.safety = libsafety_py.libsafety
+    self.safety.set_safety_hooks(CarParams.SafetyModel.mazda, 1)
+    self.safety.init_tests()
+
+  def _pcm_status_msg(self, enable):
+    values = {"ACC_ACTIVE": enable, "BRAKE_ON": 0}
+    return self.packer.make_can_msg_safety("PEDALS", 0, values)
+
```
[the remark this answers is not in the document]
Comment (yummydirtx) [reply opendbc/safety/tests/test_mazda.py]: Great call. I implemented longitudinal accel-limit coverage for Mazda long mode by inheriting from LongitudinalAccelSafetyTest, adding MAX/MIN/INACTIVE accel constants, and implementing _accel_msg() for CRZ_INFO.ACCEL_CMD. I also overrode the limit test for Mazda to use integer boundary sweeps since ACCEL_CMD is a raw integer command field here. This avoids float packing artifacts while still verifying tx-hook limits thoroughly. The Mazda safety test module now passes locally.

File: opendbc/car/mazda/longitudinal.py [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+                                 hold_latched: bool = False, crz_hold_latched: bool = False,
+                                 crz_hold_passive: bool = False,
+                                 crz_resume_active: bool = False,
+                                 crz_info_resume_unlatching: bool = False,
+                                 v_ego: float = 0.0) -> list[CanData]:
+  if crz_ctrl_hold_request is None:
+    crz_ctrl_hold_request = hold_request
+
+  return [
+    CanData(CRZ_INFO_ADDR, build_crz_info(accel, counter, long_active, hold_request, v_ego,
```
[the remark this answers is not in the document]
Comment (yummydirtx) [reply opendbc/car/mazda/longitudinal.py:197]: Agreed on the API cleanup point. I removed the unused standstill parameter from create_longitudinal_messages() and updated the caller accordingly. I intentionally did not fold standstill into hold_request behavior, since that would change stop/hold semantics beyond cleanup and the current state machine already drives CRZ_INFO/CRZ_CTRL hold behavior through explicit hold_request/crz_ctrl_hold_request inputs.

**Comment by yummydirtx** [comment]
Marked this as a draft while I prepare a submission to commaai/opendbc as suggested by sunnyhaibin
